src/template/transaction/executor.ts

```diff
@@ -0,0 +1,822 @@
+  interpreter: ITxTemplateInterpreter,
+  ctx: TxTemplateContext,
+  ins: z.infer<typeof ChangeInstruction>
+) {
+  ctx.log(`Begin ChangeInstruction: ${JSONBigInt.stringify(ins)}`);
+  const token = getVariable<string | undefined>(ins.token, ctx.vars, ChangeInstruction.shape.token);
+  const address =
+    getVariable<string | undefined>(ins.address, ctx.vars, ChangeInstruction.shape.address) ??
+    (await interpreter.getChangeAddress(ctx));
+  const timelock = getVariable<number | undefined>(
+    ins.timelock,
+    ctx.vars,
+    ChangeInstruction.shape.timelock
+  );
+  ctx.log(`address(${address}) timelock(${timelock}) token(${token})`);
+
+  const tokensToCheck: string[] = [];
+  if (token) {
+    tokensToCheck.push(token);
+  } else {
+    // Check HTR and all tokens on the transaction
+    tokensToCheck.push(NATIVE_TOKEN_UID);
+    for (const tk of ctx.tokens) {
+      tokensToCheck.push(tk);
+    }
+  }
+
+  const script = createOutputScriptFromAddress(address, interpreter.getNetwork());
+
+  for (const tokenUid of tokensToCheck) {
+    ctx.log(`Checking change tx for token ${tokenUid}`);
+    const balance = ctx.balance.getTokenBalance(tokenUid);
+    const tokenData = ctx.addToken(tokenUid);
+    if (balance.tokens > 0) {
+      const value = balance.tokens;
+      ctx.log(`Adding a change output for ${value} / ${tokenUid}`);
+      // Need to create a token output
+      // Add balance to the ctx.balance
+      ctx.balance.addOutput(value, tokenUid);
+
+      // Creates an output with the value of the outstanding balance
+      const output = new Output(value, script, { timelock, tokenData });
+      ctx.addOutput(-1, output);
+    }
+
+    if (balance.mint_authorities > 0) {
+      const count = balance.mint_authorities;
+      ctx.log(`Adding ${count} mint authority change outputs / ${tokenUid}`);
+      // Need to create a token output
+      // Add balance to the ctx.balance
+      ctx.balance.addOutputAuthority(count, tokenUid, 'mint');
+
+      // Creates an output with the value of the outstanding balance
+      const output = new Output(TOKEN_MINT_MASK, script, {
+        timelock,
+        tokenData: tokenData | TOKEN_AUTHORITY_MASK,
+      });
+      ctx.addOutput(-1, ...Array(count).fill(output));
+    }
+
+    if (balance.melt_authorities > 0) {
+      const count = balance.mint_authorities;
+      ctx.log(`Adding ${count} melt authority change outputs / ${tokenUid}`);
+      // Need to create a token output
+      // Add balance to the ctx.balance
+      ctx.balance.addOutputAuthority(count, tokenUid, 'melt');
+
+      // Creates an output with the value of the outstanding balance
+      const output = new Output(TOKEN_MELT_MASK, script, {
+        timelock,
+        tokenData: tokenData | TOKEN_AUTHORITY_MASK,
+      });
+      ctx.addOutput(-1, ...Array(count).fill(output));
+    }
+  }
+}
+
+/**
+ * Execution for CompleteTxInstruction
+ */
+export async function execCompleteTxInstruction(
```

Comment:
Maybe you could create a method to fill inputs and use the method above and the new one. I feel that you are duplicating exactly the same code as above only because this 'complete' method fills also inputs.

Maybe just reuse the methods above to select utxos for authority and tokens.

What do you think?

Reply:
This is part of what I said about allowing some code duplication on the executors in this first PR.
Later we can abstract based on usage.